### PR TITLE
[Refactor][Ops] Keep pad/slice adaptation on the kernel side

### LINKED ISSUE #1818: [REFACTOR][OPS] stop the op layer from reading kernel-internal state
### Symptom / Motivation

The op layer depends on kernel-internal state to interpret a result. It happens in two forms that share one direction — information flowing from the implementation up into the op:

- A kernel returns a padded shape or a surplus workspace buffer, and the op trims it. `tileops/ops/norm/rms_norm.py` receives `(M, N_padded)` and slices to N. `tileops/ops/engram.py` receives 7 buffers, drops `dvhat_buf`, and returns 6 semantic gradients.
- The op reads a private attribute off the kernel object to learn the output dtype.

This is a layering defect on the repo's own terms: the op layer should not know about `N_padded` or `dvhat_buf`. It also blocks any second kernel implementation, which would have to reproduce padding and workspace conventions it cannot see.

### Root Cause Analysis

11 of 174 kernel-map slots return a raw result the op has to post-process. Measured list, grouped by the op file that does the trimming:

| slot | op file |
| --- | --- |
| `rms_norm` | `tileops/ops/norm/rms_norm.py` |
| `fused_add_layer_norm` | `tileops/ops/norm/fused_add_layer_norm.py` |
| `fused_add_rms_norm` | `tileops/ops/norm/fused_add_rms_norm.py` |
| `group_norm`, `group_norm_no_affine` | `tileops/ops/norm/group_norm.py` |
| `group_norm`, `instance_norm_no_affine` | `tileops/ops/norm/instance_norm.py` |
| `softmax_fwd` | `tileops/ops/reduction/softmax.py` |
| `cumulative_fwd` | `tileops/ops/reduction/cumulative.py` |
| `engram_gate_conv_fwd`, `engram_gate_conv_bwd` | `tileops/ops/engram.py` |
| `engram_decode` | `tileops/ops/engram_decode.py` |

Separately, `tileops/ops/elementwise/_base.py` calls `getattr(kernel, "_fp8_output_dtype", None)` at lines 552, 620, 780, 886 and 1057 — five reads of a private attribute. The manifest already answers this question: 188 of 188 ops declare their output dtype under `signature.outputs.<name>.dtype`. The spec holds the authoritative value and the code asks the implementation instead.

### Related Files

- `tileops/ops/norm/rms_norm.py`, `fused_add_layer_norm.py`, `fused_add_rms_norm.py`, `group_norm.py`, `instance_norm.py`
- `tileops/ops/reduction/softmax.py`, `cumulative.py`
- `tileops/ops/engram.py`, `tileops/ops/engram_decode.py`
- `tileops/ops/elementwise/_base.py`
- `tileops/kernels/norm/`, `tileops/kernels/reduction/`, `tileops/kernels/engram/`

## Goal

A kernel hands back exactly the semantic result, and the op derives output dtype from the manifest — so no op file reads or repairs kernel-internal state.

## Plan

1. For each of the 11 slots, move the pad/slice and the workspace-buffer drop from the op into a thin wrapper on the kernel side under `tileops/kernels/`. Do not change the `prim_func` — writing unaligned output directly can cost memory throughput; the goal is to relocate the adaptation, not delete it.
2. Derive elementwise output dtype from the manifest `signature.outputs` entry, and delete the five `_fp8_output_dtype` reads.
3. Re-run the affected benchmarks against the pre-change numbers to show the relocation is free.

## Constraints

- Output values, shapes and dtypes are unchanged for every public op — this is a relocation of existing logic, not a behavior change.
- No manifest content changes: the output-dtype values being read are already present.
- No performance regression on norm, softmax, cumulative or engram.
- Large human-driven layering refactor. Do not split the op-side and kernel-side halves of a single slot across PRs; they are one change.

## Acceptance Criteria

- [ ] Modified files pass unit tests
- [ ] `grep -rn '_fp8_output_dtype' tileops/ops/` returns nothing
- [ ] None of the 9 op files listed above slices padded kernel output or selects a subset of a returned tuple
- [ ] `pytest tests/ -m smoke` green, and full-tier tests green for every op touched
- [ ] Benchmark comparison for norm / softmax / cumulative / engram posted in the PR, showing no regression

### PR DESCRIPTION
**Pull request by lcy-seso**
Closes #1818

## Summary

The op layer no longer reads kernel-internal state (#1818), plus the duplication cleanup folded in on top. Behaviour-preserving throughout: same outputs, shapes and dtypes, no manifest content changes.

- 11 slots returned padded or workspace buffers the op then trimmed; that adaptation moved into kernel-side wrappers, no `prim_func` touched
- Elementwise output dtype comes from the manifest instead of `getattr(kernel, "_fp8_output_dtype")`
- One dtype-expression grammar, was 3 regexes that had already drifted apart
- One manifest-entry lookup, was 2
- One `op_base` FIXME block, was 3 for a single degradation
- One tiled-autotune sweep, was 3 differing only in probe dtype
- One base per max-pool dimension, was value/indices pairs 98% identical
- One GQA search space, was 12 byte-identical copies
- One parameterised convolution search space, was 9 near-copies
- Two more shared oracles in `tileops/testing/`, was a verbatim copy in both the test and the benchmark — the mamba docstrings had already disagreed
- Convolution rows as a `ConvCase`, was 8 flattened params across 6 functions

48 files, +955/−1601. Collected test nodes unchanged at 5000.

## Reviewer notes

- `GroupNormKernel` affine args became optional with cached identity buffers. Without it, relocating the padding cost 25-35% on unaligned shapes; the first benchmark pass caught it.
- Deliberately **not** unified: `validate_manifest.py` keeps its own token splitter (it must see empty tokens so a trailing `|` still errors); max-pool dimensions stay separate (different constructor signatures are the calling contract); the six convolution benchmarks stay separate (the validator matches `ManifestBenchmark(<op name>, …)` statically).
- Out of scope: the `deltanet_fwd` / `gated_deltanet_fwd` autotune pair — nothing covers `tune=True` there and #1821 is a live bug in the same area. Tracked as #1822.

## Test plan

- [x] Modified files pass unit tests
- [x] `pytest tests/ -m smoke` 2626 passed; full tier on every touched op 3736 passed
- [x] `grep -rn '_fp8_output_dtype' tileops/ops/` empty; no op file trims kernel output
- [x] Max-pool checked against `torch.nn.functional` for 1d/2d/3d including indices (`torch.equal`, not a tolerance)
- [x] `bench_convolution` collects the same 57 ids as main, smoke subset unchanged at 10
- [x] `validate_manifest.py` clean, its 121 tests pass
- [x] Search spaces unchanged: identical config lists for 3 dtypes across all 5 convolution spaces, identical ordering for the 54 GQA entries
- [x] `tune_by_forward` verified with a stub kernel — every candidate tried, fastest wins, empty list falls back to `default_config`

## Benchmark

H200, CUDA 12.8, PyTorch 2.9.1+cu128, TileLang 0.1.11. 84 workloads measured at `upstream/main` and at HEAD.

| Family | Rows | Mean | Best | Worst |
| --- | ---: | ---: | ---: | ---: |
| norm | 49 | -0.01% | -0.76% | +0.71% |
| softmax | 20 | +0.03% | -0.28% | +0.44% |
| cumulative | 6 | -0.12% | -0.23% | +0.00% |
| engram | 9 | +0.07% | -0.24% | +0.85% |

```
CUDA_VISIBLE_DEVICES=2 pytest -q benchmarks/ops/bench_norm.py benchmarks/ops/bench_group_norm.py \
  benchmarks/ops/bench_instance_norm.py benchmarks/ops/bench_softmax.py \
  benchmarks/ops/bench_cumulative.py benchmarks/ops/bench_engram.py
```

## Regression

Largest absolute movement 0.85%, consistent with timing noise. Pool and convolution benchmarks pass at HEAD (48 and 57).

### CODE REVIEW

**Review by Ibuki-wind — DISMISSED** [review]

File: tileops/kernels/reduction/_primitives.py [hunk 1]
```diff
@@ -198,6 +200,44 @@ def compute_tile_n(
...
+            kernel.forward(*probe_inputs)
+        end.record()
+        torch.cuda.synchronize()
+        elapsed = start.elapsed_time(end) / rep
+
+        if elapsed < best_time:
+            best_time, best_config = elapsed, cfg
+
+    kernel.config = best_config
+    print(f'Best config: {kernel.config}')
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/reduction/_primitives.py:239]: Correct about the behaviour, but not changing it here. The three copies this replaces had no per-candidate guard either, so adding one would make a behaviour-preserving refactor change behaviour.

#1821 tracks it and argues for the other fix: derive `tile_n` per `block_m` so every emitted candidate is buildable, rather than skipping the ones that are not. A candidate list that cannot be compiled is wrong regardless of who consumes it, and that issue explicitly rules out widening the exception handler alone.

**Review by Ibuki-wind — APPROVED** [review]
